kbase-extension/static/kbase/js/common/cellComponents/paramsWidget.js

```diff
@@ -0,0 +1,551 @@
+                            else if (original.text_options && original.text_options.is_output_name) {
+                                isParameter = false;
+                            } 
+
+                            //all other cases should be a param element
+                            else {
+                                isParameter = true;
+                            }
+                        } 
+                        
+                        return isParameter;
+
+                    }).map(function (id) {
+                        return params.specs[id];
+                    }));
+
+                return Promise.resolve()
+                    .then(function () {
+                        if (!parameterParams.layout.length) {
+                            // TODO: should be own node
+                            ui.getElement('parameters-area').classList.add('hidden');
+                        } else {
+                            places.parameterFields.innerHTML = parameterParams.content;
+
+                            return Promise.all(parameterParams.layout.map(function (parameterId) {
+                                const spec = parameterParams.paramMap[parameterId];
+                                try {
+                                    return makeFieldWidget(appSpec, spec, initialParams[spec.id])
+                                        .then(function (widget) {
+                                            widgets.push(widget);
+
+                                            return widget.start({
+                                                node: document.getElementById(parameterParams.view[spec.id].id)
+                                            });
+                                        });
+                                } catch (ex) {
+                                    console.error('Error making input field widget', ex);
+                                    var errorDisplay = div({ style: { border: '1px red solid' } }, [
+                                        ex.message
+                                    ]);
+                                    document.getElementById(parameterParams.view[spec.id].id).innerHTML = errorDisplay;
+                                }
+                            }));
+                        }
+                    })
+                    .then(function () {
+                        renderAdvanced();
+                    });
+            });
+        }
+
+        function start(arg) {
+            return Promise.try(function () {
+                // send parent the ready message
+
+                doAttach(arg.node);
+
+                model.setItem('appSpec', arg.appSpec);
+                model.setItem('parameters', arg.parameters);
+
+                paramsBus.on('parameter-changed', function (message) {
+                    // Also, tell each of our inputs that a param has changed.
+                    // TODO: use the new key address and subscription
```

Comment:
this TODO was in here, would be happy to update if anyone knows what this comment is about